Update task summary (#21067)
huggingface/transformers: docs/source/en/tasks_explained.mdx

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# How 🤗 Transformers solve tasks

In [What 🤗 Transformers can do](task_summary), you learned how natural language processing (NLP), speech and audio, and computer vision problems used to be solved before the Transformer was introduced and some important applications of these problems. 

This page will dig deeper into how Transformer models solve these tasks, and explain what's happening under the hood for each of them. Each Transformer model is a bit different - some may implement certain tricks or techniques - but the general idea is the same. So to keep things relatively brief, for each task, we'll discuss how a specific Transformer model works. Then, to use the pretrained model for a specific downstream task, you only need to add a task-specific head on top and finetune it on your data.

> The Transformer models we'll look at are among the most common ones in their field, and they'll provide you with a good understanding of how they solve a task:

Author: Since there are different ways to solve a task, my approach was to pick the most common model used for a task and then explain how it works. Do you think this leaves out other ways a task can be solved? For example, for image classification, I chose ViT since it is so popular, but this doesn't explain how CNNs work. Likewise, you can also do audio classification by turning audio into an image and classifying that. 

Would adding a disclaimer that this isn't a comprehensive guide covering all the ways a task is solved help?

Would love some feedback on this! cc @sgugger @MKhalusova

Reviewer: For computer vision, I think it would be worthwhile to explain the two approaches (CNN and Transformer) as they are pretty different, and both are used a lot.

For the other example in audio, you can add a note, but it's just a clever way of changing the modality from audio to vision, not a new model.

Reviewer 2: @sgugger Are there purely CNN models available via transformers library? Because this doc is about how Transformers library approaches the way to solve tasks. There are hybrid models (like DETR), so a short mention of what CNN part does is definitely useful, but if there are no pure CNN models available via Transformers, perhaps a reference to another resource would suffice?

Author: Thanks for the pointer, I'll add an explanation for CNN then!

> Are there purely CNN models available via transformers library?

Yes, we have several like ConvNeXT, ResNet, MobileNets, and I think BiT as well.

- [GLPN](model_doc/glpn) for depth estimation
- [BERT](model_doc/bert) for NLP tasks like text classification, token classification and question answering that use an encoder
- [GPT2](model_doc/gpt2) for NLP tasks like text generation that use a decoder
- [BART](model_doc/bart) for NLP tasks like summarization and translation that use an encoder-decoder

<Tip>

Before you go further, it is good to have a basic understanding of the original Transformer architecture. Knowing how encoders, decoders, and attention work would really help. If you're just getting started or need a refresher, check out our [course](https://huggingface.co/course/chapter1/4?fw=pt) for more information about how Transformers work! 

</Tip>

## Speech and audio

[wav2vec2](model_doc/wav2vec2) is a self-supervised model pretrained on unlabeled speech data and finetuned on labeled data for audio classification and automatic speech recognition. This model has four main components:

1. A *feature encoder* takes the raw waveform, normalizes it to zero mean and unit variance, and converts it into a sequence of feature vectors that are each 20ms long.

> 2. Next, the feature vectors are passed to a *quantization module*. The quantizer's goal is to learn discrete speech units. Audio waveforms are continuous by nature, so they can't be broken up into separate units like a sequence of text can be split into words. So what are speech units? Each speech unit is a speech representation sampled from a group of *codebooks*, and each codebook contains a set vocabulary. wav2vec2 has two codebooks with 320 codewords in each codebook for a total of 102,400 speech units.

Author: I tried to explain this a little better since I think it's important for the user to understand where the input is coming from. Maybe one of our audio experts can help give a clearer explanation 😅

1. Use a modern CNN, like ConvNeXT, which relies on convolutional layers but adopts new network designs.
2. Use a Transformer architecture to split an image into patches and process the patches as a sequence.

<Tip>

A third approach is a hybrid Transformer with convolutions (for example, [Convolutional Vision Transformer](model_doc/cvt), or [LeViT](model_doc/levit)). We won't discuss those here because they just combine the two approaches we examine here.

</Tip>

ViT and ConvNeXT are commonly used for image classification, but for other vision tasks like object detection, segmentation, and depth estimation, we'll take a look at DETR and GLPN, respectively; these models are better suited for those tasks.

### Image classification

ViT and ConvNeXT are both used for image classification; the main difference is that ViT uses a Transformer architecture while ConvNeXT uses convolutions.

#### Transformer

[ViT](model_doc/vit) replaces convolutions entirely with a pure Transformer architecture. If you're familiar with the original Transformer, then you're already most of the way toward understanding ViT.

The main change ViT introduced was in how images are processed:

> 1. An image is split into square patches. If you had a 224x224 pixel image, you could split it into 256 14x14 image patches. Just like how text is tokenized into words, an image is "tokenized" into a sequence of patches. Each patch is flattened by concatenating the height and width pixel values into a single vector. For a 3D image, the height, width, and number of *channels* are flattened.

Reviewer 2: I think you can shorten this paragraph by only mentioning that "Each patch is then flattened into a single vector." This is likely enough to know here.

2. A *learnable embedding* - a special `[CLS]` token - is added to the beginning of the patch embeddings just like in the BERT model. The encoder output with the `[CLS]` token is used as the input to the attached classification head; other outputs are ignored. This token helps the model learn how to encode a representation of the image.

3. The last thing to add to the patch and learnable embedding is the *position embedding* because the model doesn't know how the image patches are ordered. The position embeddings are also learnable and have the same size as the patch embeddings. Finally, all of the embeddings are passed to the Transformer encoder.

4. The output, specifically only the output with the `[CLS]` token, is passed to a multilayer perceptron head (MLP). ViT's pretraining objective is simply classification. Like other classification heads, the MLP head converts the output into logits and calculates the cross-entropy loss to find the most likely label.

Ready to try your hand at image classification? Check out our complete [image classification guide](tasks/image_classification) to learn how to finetune ViT and use it for inference!

#### ConvNet

> [ConvNeXT](model_doc/convnext) is a CNN architecture that adopts new and modern network designs to improve performance. However, convolutions are still at the core of the model. From a high-level perspective, a [convolution](glossary#convolution) looks at a small window of the image pixels and computes some features from it such as a certain texture or curvature of a line. The distance the convolution travels is known as the *stride*. You can feed this output to another convolutional layer, and with each successive layer, the network learns more complex and abstract things like hotdogs or rockets. Between convolutional layers, it is common to add a pooling layer to reduce dimensionality and make the model more robust to variations of a feature's position. This is the basic idea behind CNNs.

Reviewer 2: Perhaps, just like with Transformers, we should explicitly say what level of familiarity with CNNs we expect from a reader, and provide some link with basic concepts?

From here, ConvNeXT modernizes the CNN in five ways:

> 1. Change the number of blocks in each stage and "patchify" an image with a larger stride and corresponding kernel size. The non-overlapping sliding window makes this patchifying strategy similar to how ViT splits an image into patches.

Reviewer 2: You explain what stride is but this is the first mention of kernel. I'm not sure if it's better to add a definition of kernel, or to remove mention of both for simplicity.

Author: Added a brief explanation of kernel so we can use this vocabulary and especially since it comes up a lot in computer vision

> 2. Replace the typical 3x3 convolutional layer in the bottleneck block with *depthwise convolution*, which applies a convolution to each input channel separately and then stacks them back together at the end. This widens the network width which improves performance.

Reviewer 2: This is the first mention of the bottleneck block, should we explain it first?

Author: Switched the order around so what the bottleneck layer does is explained first before we get into the depthwise convolutions

2. BERT is pretrained with two objectives: masked language modeling and next sentence prediction. In masked language modeling, some percentage of the input tokens are randomly masked, and the model needs to predict these. This solves the issue of deep bidirectionality, where the model could cheat and see all the words and "predict" the next word. The final hidden states of the predicted mask tokens are passed to a feedforward network with a softmax over the vocabulary to predict the masked word.

    The second pretraining object is next sentence prediction. The model must predict whether sentence B is the next sentence after sentence A. Half of the time sentence B is the next sentence and the other half of the time, sentence B is a random sentence. The prediction, whether it is the next sentence or not, is passed to a feedforward network with a softmax over the two classes (`IsNext` and `NotNext`).

> 3. The input embeddings are passed through multiple encoder layers to output some final hidden states. To use the pretrained model for classification, add a sequence classification head on top of the base BERT model. The sequence classification head is just a linear layer that accepts the final hidden states and does a linear transformation to convert them into logits. The cross-entropy loss is then calculated between the logits and target to find the most likely label.

Reviewer: This is not text classification but masked language modeling or BERT pretraining. Like you did for Wav2Vec2, there should be a section after the pretraining one that details how to get to text classification from the pretrained model.

2. The encoder's output is passed to the decoder, which must predict the masked tokens and any uncorrupted tokens from the encoder's output. This gives additional context to help the decoder restore the original text. The output from the decoder is passed to a language modeling head, which does a linear transformation to convert the hidden states into logits. The cross-entropy loss is then calculated between the logits and the label, which is just the token shifted to the right.

> Ready to try your hand at summarization? Check out our complete [summarization guide](tasks/summarization) to learn how to finetune T5 and use it for inference!

Reviewer: Since this is using T5, maybe mention it in the intro with BART?

-->

# How 🤗 Transformers solve tasks

In [What 🤗 Transformers can do](task_summary), you learned about natural language processing (NLP), speech and audio, computer vision tasks, and some important applications of them. This page will look closely at how models solve these tasks and explain what's happening under the hood. There are many ways to solve a given task, some models may implement certain techniques or even approach the task from a new angle, but for Transformer models, the general idea is the same. Owing to its flexible architecture, most models are a variant of an encoder, decoder, or encoder-decoder structure. In addition to Transformer models, our library also has several convolutional neural networks (CNNs), which are still used today for computer vision tasks. We'll also explain how a modern CNN works.

To explain how tasks are solved, we'll walk through how models are pretrained and equipped with task-specific heads to output useful predictions.

- [wav2vec2](model_doc/wav2vec2) for audio classification and automatic speech recognition (ASR)
- [Vision Transformer (ViT)](model_doc/vit) and [ConvNeXT](model_doc/convnext) for image classification

> - [DETR](model_doc/detr) for image segmentation and object detection
> - [GLPN](model_doc/glpn) for depth estimation

Reviewer 3: ```suggestion
- [DETR](model_doc/detr) for object detection
- [Mask2Former](model_doc/mask2former) for image segmentation
- [GLPN](model_doc/glpn) for depth estimation
```

Author: Replaced DETR for image segmentation with Mask2Former 👍

3. About half of the feature vectors are randomly masked, and the masked positions are replaced by the same trained feature vector. The masked feature vector is fed to a *context network*. The context network is a Transformer encoder, but unlike the original Transformer, wav2vec2 adds a relative positional embedding with another convolutional layer.

4. The pretraining objective of the context network is to learn the speech representation of the masked positions by solving a *contrastive task*. The model has to predict the true quantized speech representation from a set of false ones. The contrastive loss encourages the model to maximize the similarity between the context vector and the quantized speech unit (the target label). The output is the encoder's hidden states.

Now that wav2vec2 is pretrained, you can finetune it on your data for audio classification or automatic speech recognition!

### Audio classification

> To use the pretrained model for audio classification, add a sequence classification head on top of the base wav2vec2 model. The classification head is a linear layer that accepts the encoder's hidden states and transforms them into logits. The cross-entropy loss is calculated between the logits and target to find the most likely class.

Reviewer 4: This is probably where I'd add a bit more detail on audio classification and how the W2V2 model works in this case (the structure is good, IMO you can just expand each sentence with an extra level of detail)

> To use the pretrained model for audio classification, add a sequence classification head on top of the base wav2vec2 model. The classification head is a linear layer that accepts the encoder's hidden states and transforms them into logits. The cross-entropy loss is calculated between the logits and target to find the most likely class.

Author: I think it'll be ok since this is part of a two-part series all about tasks. The first [doc](https://huggingface.co/docs/transformers/task_summary) explains what the task is, and this one is more focused on how the task is solved. 🙂

Ready to try your hand at audio classification? Check out our complete [audio classification guide](tasks/audio_classification) to learn how to finetune wav2vec2 and use it for inference!

### Automatic speech recognition

To use the pretrained model for automatic speech recognition, add a language modeling head on top of the base wav2vec2 model for [connectionist temporal classification (CTC)](glossary#connectionist-temporal-classification-ctc). The language modeling head is a linear layer that accepts the encoder's hidden states and transforms them into logits. Each logit represents a token class (the number of tokens comes from the task vocabulary). The CTC loss is calculated between the logits and target to find the most likely token, which is decoded into a word.

Ready to try your hand at automatic speech recognition? Check out our complete [automatic speech recognition guide](tasks/asr) to learn how to finetune wav2vec2 and use it for inference!

## Computer vision

There are two ways to approach computer vision tasks:

> 1. Use a modern CNN, like ConvNeXT, which relies on convolutional layers but adopts modern network designs.

Reviewer 3: ```suggestion
1. Use a modern CNN, like [ConvNeXT](...), which relies on convolutional layers but adopts modern network designs.
```
Would be great to link to their docs

> 2. Use a Transformer architecture to split an image into patches and process the patches sequentially.

Reviewer 3: Transformers don't process patches sequentially, but rather in parallel

<Tip>

A third approach mixes Transformers with convolutions (for example, [Convolutional Vision Transformer](model_doc/cvt), or [LeViT](model_doc/levit)). We won't discuss those because they just combine the two approaches we examine here.

</Tip>

ViT and ConvNeXT are commonly used for image classification, but for other vision tasks like object detection, segmentation, and depth estimation, we'll look at DETR and GLPN, respectively; these models are better suited for those tasks.

### Image classification

ViT and ConvNeXT are both used for image classification; the main difference is that ViT uses a Transformer architecture while ConvNeXT uses convolutions.

#### Transformer

[ViT](model_doc/vit) replaces convolutions entirely with a pure Transformer architecture. If you're familiar with the original Transformer, then you're already most of the way toward understanding ViT.

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/model_doc/vit_architecture.jpg"/>
</div>

The main change ViT introduced was in how images are processed:

> 1. An image is split into square patches. If you had a 224x224 pixel image, you could split it into 256 14x14 image patches. Just like how text is tokenized into words, an image is "tokenized" into a sequence of patches. Each patch is then flattened into a single vector.

Reviewer 3: ```suggestion
1. An image is split into square non-overlapping patches. If you had a 224x224 pixel image, you could split it into 196 16x16 image patches. Just like how text is tokenized into words, an image is "tokenized" into a sequence of patches. Each patch is then flattened into a single vector.
```

>     The pixel values are linearly transformed into *patch embeddings* by a 2D convolution to create the proper input dimensions. Similar image patches are mapped to similar patch embeddings.

Reviewer 3: ```suggestion
    The patches are linearly transformed into *patch embeddings* by a 2D convolution to create the proper input dimensions.
```

Ready to try your hand at automatic speech recognition? Check out our complete [automatic speech recognition guide](tasks/asr) to learn how to finetune Wav2Vec2 and use it for inference!

## Computer vision

There are two ways to approach computer vision tasks:

1. Use a modern CNN, like [ConvNeXT](model_doc/convnext), which relies on convolutional layers but adopts modern network designs.
2. Use a Transformer architecture to split an image into a sequence of patches and process them in parallel.

<Tip>

A third approach mixes Transformers with convolutions (for example, [Convolutional Vision Transformer](model_doc/cvt), or [LeViT](model_doc/levit)). We won't discuss those because they just combine the two approaches we examine here.

</Tip>

ViT and ConvNeXT are commonly used for image classification, but for other vision tasks like object detection, segmentation, and depth estimation, we'll look at DETR and GLPN, respectively; these models are better suited for those tasks.

### Image classification

> ViT and ConvNeXT are both used for image classification; the main difference is that ViT uses a Transformer architecture while ConvNeXT uses convolutions.

Reviewer 3: ```suggestion
ViT and ConvNeXT are both used for image classification; the main difference is that ViT uses an attention mechanism while ConvNeXT uses convolutions.
```

#### Transformer

[ViT](model_doc/vit) replaces convolutions entirely with a pure Transformer architecture. If you're familiar with the original Transformer, then you're already most of the way toward understanding ViT.

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/model_doc/vit_architecture.jpg"/>
</div>

The main change ViT introduced was in how images are processed:

> 1. An image is split into square non-overlapping patches. If you had a 224x224 pixel image, you could split it into 196 16x16 image patches. Just like how text is tokenized into words, an image is "tokenized" into a sequence of patches. Each patch is then flattened into a single vector.
> 
>     The patches are linearly transformed into *patch embeddings* by a 2D convolution to create the proper input dimensions.

Reviewer 3: ```suggestion
1. An image is split into square non-overlapping patches. If you had a 224x224 pixel image, you could split it into 196 16x16 image patches. Just like how text is tokenized into words, an image is "tokenized" into a sequence of patches. Next, the patches are linearly transformed into *patch embeddings* to create the proper input dimensions (which for a base Transformer is 768 values for each embedding vector). In practice, patchifying an image is done using a convolutional 2D layer.
```

3. The last thing to add to the patch and learnable embeddings are the *position embeddings* because the model doesn't know how the image patches are ordered. The position embeddings are also learnable and have the same size as the patch embeddings. Finally, all of the embeddings are passed to the Transformer encoder.

4. The output, specifically only the output with the `[CLS]` token, is passed to a multilayer perceptron head (MLP). ViT's pretraining objective is simply classification. Like other classification heads, the MLP head converts the output into logits and calculates the cross-entropy loss to find the most likely label.

Ready to try your hand at image classification? Check out our complete [image classification guide](tasks/image_classification) to learn how to finetune ViT and use it for inference!

#### ConvNet

<Tip>

This section briefly explains convolutions, but it'd be helpful to have a basic understanding of how they change an image's shape and size. If you're unfamiliar with convolutions, check out the [Convolution Neural Networks chapter](https://github.com/fastai/fastbook/blob/master/13_convolutions.ipynb) from the fastai book!

</Tip>

> [ConvNeXT](model_doc/convnext) is a CNN architecture that adopts new and modern network designs to improve performance. However, convolutions are still at the core of the model. From a high-level perspective, a [convolution](glossary#convolution) is an operation where a smaller matrix (*kernel*) is multiplied by a small window of the image pixels. It computes some features from it, such as a particular texture or curvature of a line. Then it slides over to the next window of pixels; the distance the convolution travels is known as the *stride*. You can feed this output to another convolutional layer, and with each successive layer, the network learns more complex and abstract things like hotdogs or rockets. Between convolutional layers, it is common to add a pooling layer to reduce dimensionality and make the model more robust to variations of a feature's position. This is the basic idea behind CNNs.

Reviewer 3: This repo has some very nice visualizations of convolutions: https://github.com/vdumoulin/conv_arithmetic#convolution-animations

Maybe we can include a gif here

Author: Great idea!

### Image segmentation

Mask2Former is a universal architecture for solving all types of segmentation tasks. Traditional segmentation models classifies all the pixels in an image, but Mask2Former frames the task as a *mask classification* problem. Mask classification groups pixels into *N* segments, and predicting *N* masks and their corresponding class label.

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/mask2former_architecture.png"/>
</div>

There are three main components to Mask2Former:

1. A standard [ResNet101](model_doc/resnet) backbone accepts an image and creates a low-resolution image feature map with 3 consecutive 3x3 convolutions.

2. The feature map is passed to a *pixel decoder* which gradually upsamples the low-resolution features into high-resolution per-pixel embeddings. The pixel decoder actually generates multi-scale features (contains both low- and high-resolution features) with resolutions 1/32, 1/16, and 1/8th of the original image.

> 3. Each of these feature maps of differing scales are fed successively to one Transformer decoder layer at a time in order to capture small objects from the high-resolution features. The key to Mask2Former is the *masked attention* mechanism in the decoder. Unlike cross-attention which can attend to the entire image, masked attention only focuses on a certain area of the image. This is faster and leads to better performance because the local features of an image are enough for the model to learn from.

Reviewer 2: ```suggestion
3. Each of these feature maps of differing scales is fed successively to one Transformer decoder layer at a time in order to capture small objects from the high-resolution features. The key to Mask2Former is the *masked attention* mechanism in the decoder. Unlike cross-attention which can attend to the entire image, masked attention only focuses on a certain area of the image. This is faster and leads to better performance because the local features of an image are enough for the model to learn from.
```

Ready to try your hand at automatic speech recognition? Check out our complete [automatic speech recognition guide](tasks/asr) to learn how to finetune Wav2Vec2 and use it for inference!

## Computer vision

There are two ways to approach computer vision tasks:

1. Split an image into a sequence of patches and process them in parallel with a Transformer.
2. Use a modern CNN, like [ConvNeXT](model_doc/convnext), which relies on convolutional layers but adopts modern network designs.

<Tip>

A third approach mixes Transformers with convolutions (for example, [Convolutional Vision Transformer](model_doc/cvt) or [LeViT](model_doc/levit)). We won't discuss those because they just combine the two approaches we examine here.

</Tip>

ViT and ConvNeXT are commonly used for image classification, but for other vision tasks like object detection, segmentation, and depth estimation, we'll look at DETR, Mask2Former and GLPN, respectively; these models are better suited for those tasks.

### Image classification

> ViT and ConvNeXT are both used for image classification; the main difference is that ViT uses an attention mechanism while ConvNeXT uses convolutions.

Reviewer 3: ```suggestion
ViT and ConvNeXT can both be used for image classification; the main difference is that ViT uses an attention mechanism while ConvNeXT uses convolutions.
```

#### Transformer

[ViT](model_doc/vit) replaces convolutions entirely with a pure Transformer architecture. If you're familiar with the original Transformer, then you're already most of the way toward understanding ViT.

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/model_doc/vit_architecture.jpg"/>
</div>

> The main change ViT introduced was in how images are processed:

Reviewer 3: ```suggestion
The main change ViT introduced was in how images are fed to a Transformer:
```

> 1. An image is split into square non-overlapping patches. If you had a 224x224 pixel image, you could split it into 196 16x16 image patches. Just like how text is tokenized into words, an image is "tokenized" into a sequence of patches. Each patch is then flattened into a single vector. Next, the patches are linearly transformed into *patch embeddings* to create the proper input dimensions (which for a base Transformer is 768 values for each embedding vector). In practice, patchifying an image is done using a convolutional 2D layer.

Reviewer 3: ```suggestion
1. **Patch embeddings**: An image is split into square non-overlapping patches, each of which gets turned into a vector (embedding). If you had a 224x224 pixel image, you could split it into 196 16x16 image patches. Just like how text is tokenized into words, an image is "tokenized" into a sequence of patches. The patch embeddings are created using a convolutional 2D layer to create the proper input dimensions (which for a base Transformer is 768 values for each patch embedding).
```
Sorry what I said above is slightly wrong, patches aren't flattened

4. ViT has a global receptive field which means it can see more of an image at once thanks to its attention mechanism. ConvNeXT attempts to replicate this effect by increasing the kernel size to 7x7.

5. ConvNeXT also makes several layer design changes that imitate Transformer models. There are fewer activation and normalization layers,  the activation function is switched to GELU instead of ReLU, and it uses LayerNorm instead of BatchNorm.

The output from the convolution blocks is passed to a classification head which converts the outputs into logits and calculates the cross-entropy loss to find the most likely label.

### Object detection

[DETR](model_doc/detr), *DEtection TRansformer*, is an end-to-end object detection model that combines a CNN with a Transformer encoder-decoder.

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/detr_architecture.png"/>
</div>

> 1. A pretrained CNN *backbone* takes an image, represented by its pixel values, and creates a low-resolution activation map of the image features. A 1x1 convolution is applied to the feature map to reduce dimensionality and it creates a new feature map with a high-level image representation. Since the Transformer is a sequential model, the feature map is flattened into a sequence of feature vectors that are combined with positional embeddings.

Reviewer 3: ```suggestion
1. A pretrained CNN *backbone* takes an image, represented by its pixel values, and creates a low-resolution feature map of it. A 1x1 convolution is applied to the feature map to reduce dimensionality and it creates a new feature map with a high-level image representation. Since the Transformer is a sequential model, the feature map is flattened into a sequence of feature vectors that are combined with positional embeddings.
```

2. The feature vectors are passed to the encoder, which learns the image representations using its attention layers. Next, the encoder hidden states are combined with *object queries* in the decoder. Object queries are learned embeddings that focus on the different regions of an image, and they're updated as they progress through each attention layer. The decoder hidden states are passed to a feedforward network that predicts the bounding box coordinates and class label for each object query, or `no object` if there isn't one.

>     DETR decodes each object query in parallel to output *N* final predictions. Unlike a typical autoregressive model that predicts one element at a time, object detection is a set prediction task (`bounding box`, `class label`) that makes *N* predictions in a single pass.

Reviewer 3: ```suggestion
DETR decodes each object query in parallel to output *N* final predictions (with N = number of queries). Unlike a typical autoregressive model that predicts one element at a time, object detection is a set prediction task (`bounding box`, `class label`) that makes *N* predictions in a single pass.
```

> 3. DETR uses *bipartite matching loss* during training to compare a fixed number of predictions with a fixed set of ground truth labels. If there are fewer ground truth labels in the set of *N* labels, then they're padded with a `no object` class. This loss function encourages DETR to find a one-to-one assignment between the predictions and ground truth labels. If either the bounding boxes or class labels aren't correct, a loss is incurred. Likewise, if DETR predicts an object that doesn't exist, it is penalized. This encourages DETR to find other objects in an image instead of focusing on one really prominent object.

Reviewer 3: ```suggestion
3. DETR uses a *bipartite matching loss* during training to compare a fixed number of predictions with a fixed set of ground truth labels. If there are fewer ground truth labels in the set of *N* labels, then they're padded with a `no object` class. This loss function encourages DETR to find a one-to-one assignment between the predictions and ground truth labels. If either the bounding boxes or class labels aren't correct, a loss is incurred. Likewise, if DETR predicts an object that doesn't exist, it is penalized. This encourages DETR to find other objects in an image instead of focusing on one really prominent object.
```

An object detection head is added on top of DETR to find the class label and the coordinates of the bounding box. There are two components to the object detection head: a linear layer to transform the decoder hidden states into logits over the class labels, and a MLP to predict the bounding box.

Ready to try your hand at object detection? Check out our complete [object detection guide](tasks/object_detection) to learn how to finetune DETR and use it for inference!

### Image segmentation

> [Mask2Former](model_doc/mask2former) is a universal architecture for solving all types of segmentation tasks. Traditional segmentation models classify all the pixels in an image, but Mask2Former frames the task as a *mask classification* problem. Mask classification groups pixels into *N* segments, and predicts *N* masks and their corresponding class label. We'll explain how Mask2Former works in this section, and then you can try finetuning SegFormer at the end.

Reviewer 3: ```suggestion
[Mask2Former](model_doc/mask2former) is a universal architecture for solving all types of image segmentation tasks. Traditional segmentation models are typically very tailored towards a particular subtask of image segmentation, like instance, semantic or panoptic segmentation. Mask2Former frames each of those tasks as a *mask classification* problem. Mask classification predicts *N* masks and their corresponding class label for a given image. We'll explain how Mask2Former works in this section, and then you can try finetuning SegFormer at the end.
```

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/mask2former_architecture.png"/>
</div>

There are three main components to Mask2Former:

> 1. A standard [ResNet101](model_doc/resnet) backbone accepts an image and creates a low-resolution image feature map from 3 consecutive 3x3 convolutions.

Reviewer 3: Mask2Former actually uses Swin Transformer as backbone.

However they also have checkpoints that leverage ResNet as backbone (with lower performance)

Author: Replaced ResNet with Swin!

    The mask predictions are generated by combining the pixel-embeddings with the final decoder hidden states. The sigmoid cross-entropy and dice loss is calculated between the logits and the ground truth mask to find the most likely mask.

Ready to try your hand at object detection? Check out our complete [image segmentation guide](tasks/semantic_segmentation) to learn how to finetune SegFormer and use it for inference!

### Depth estimation

[GLPN](model_doc/glpn), *Global-Local Path Network*, is a Transformer for depth estimation that combines a [SegFormer](model_doc/segformer) encoder with a lightweight decoder.

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/glpn_architecture.jpg"/>
</div>

1. Like ViT, an image is split into a sequence of patches, except these image patches are smaller. This is better for dense prediction tasks like segmentation or depth estimation. The image patches are transformed into patch embeddings (see the [image classification](#image-classification) section for more details about how patch embeddings are created), which are fed to the encoder.

> 2. The encoder takes the patch embedding, and image features are learned through the attention and Mix-FFN layer. The purpose of the latter is to provide positional information. At the end of the encoder block is a *patch merging* layer for creating hierarchical representations. The features of each group of neighboring patches are concatenated, and a linear layer is applied to the concatenated features to reduce the number of patches to a resolution of 1/4. This becomes the input to the next encoder block, where this whole process is repeated until you have image features with resolutions of 1/8, 1/16, and 1/32.

Reviewer 3: ```suggestion
2. The encoder takes in the patch embeddings, and passes them through several encoder blocks. Each block consists of attention and Mix-FFN layers. The purpose of the latter is to provide positional information. At the end of each encoder block is a *patch merging* layer for creating hierarchical representations. The features of each group of neighboring patches are concatenated, and a linear layer is applied to the concatenated features to reduce the number of patches to a resolution of 1/4. This becomes the input to the next encoder block, where this whole process is repeated until you have image features with resolutions of 1/8, 1/16, and 1/32.
```

> 3. A lightweight decoder takes the last feature map (1/32 scale) from the encoder and upsamples it to 1/16 scale. From here, the feature is passed into a *Selective Feature Fusion (SFF)* module, which selects and combines local and global features from an attention map for each feature and then upsamples it to 1/8th. This process is repeated until the decoded features are the same size as the original image. The output is passed through two convolution layers and then a sigmoid function to predict the depth of each pixel.

Reviewer 3: ```suggestion
3. A lightweight decoder takes the last feature map (1/32 scale) from the encoder and upsamples it to 1/16 scale. From here, the feature is passed into a *Selective Feature Fusion (SFF)* module, which selects and combines local and global features from an attention map for each feature and then upsamples it to 1/8th. This process is repeated until the decoded features are the same size as the original image. The output is passed through two convolution layers and then a sigmoid activation is applied to predict the depth of each pixel.
```
